index.html

```diff
@@ -390,15 +391,30 @@ <h4>Computation steps</h4>
                 <li id="step2F.i">Set the <code>accumulated text</code> to the empty string.</li>
                 <li id="step2F.ii">Check for <abbr title="Cascading Style Sheets">CSS</abbr> generated textual content associated with the <code>current node</code> and include it in the <code>accumulated text</code>. The <abbr title="Cascading Style Sheets">CSS</abbr> <a href="https://www.w3.org/TR/CSS2/generate.html#before-after-content"><code>:before</code> and <code>:after</code></a> pseudo elements [[!CSS2]] can provide textual content for [=element|elements=] that have a content model.
                   <ul>
-                    <li>For <code>:before</code> pseudo elements, <a class="termref">User agents</a> MUST prepend <abbr title="Cascading Style Sheets">CSS</abbr> textual content, without a space, to the textual content of the <code>current node</code>. </li>
-                    <li>For <code>:after</code> pseudo elements, <a class="termref">User agents</a> MUST append <abbr title="Cascading Style Sheets">CSS</abbr>  textual content, without a space, to the textual content of the <code>current node</code>. </li>
+                    <li>For <code>:before</code> pseudo elements, <a class="termref">User agents</a> MUST prepend <abbr title="Cascading Style Sheets">CSS</abbr> textual content to the textual content of the <code>current node</code> as follows:
+                    <ul>
+                      <li>If the pseudo element includes spacing between it and the textual content of the <code>current node</code>, <a class="termref">User agents</a> MUST prepend <abbr title="Cascading Style Sheets">CSS</abbr> textual content, with a space following, to the textual content of the <code>current node</code>. </li>
```

Comment:
Emphasis mine:

> If the pseudo element includes _spacing_ between it and the textual content of the current node

This is ambiguous and potential clarifications may be difficult to implement or costly for performance. There's no "spacing" between an element and its pseudos, but the visual spacing is depending on many factors like padding, margin, position, text-indent, borders, position of background elements or list-style-images...

Reply:
wow! It seems like this term is definitely overloaded for some folks. 

What if we said "if the pseudo element includes a space between it and the textual content of the current node..."

Maybe we could provide a code sample that demonstrates what is meant here? 

I don't want folks to overthink it but it's easy to do that with an overloaded term.

Reply:
Trying to clarify more of the ambiguity.

# "spacing" or "a space" as in what?

## whitespace characters?
- "space [character] between" (e.g. " " or other whitespace)?: TMK, there is never a character _between_ an element and its pseudo.
- space character inside the element (e.g. first child is a text node and startswith " ") "between" the visible element text and the `::before` contents?: e.g. A prepended or appended space character inside the element `<span> foo </span>? This is trimmed by other aspects of the algorithm.
- space character inside the element's pseudo?: e.g. A content value that begins with (for `::after`) or ends with (for `::before`) some whitespace character?

## Other non-whitespace-character space?
- left padding or margin of the element that has a `::before` psuedo?
- right padding or margin of the element that has an `::after` psuedo?
- left padding or margin `::after` pseudo itself?
- right padding or margin `::before` pseudo itself?
- invisible borders (e.g. `border-color: rgba(256,256,256,0);`) on either the element or pseudo?
- a pseudo with a fixed width whose background image does not fill the entire width (such as and image bullet in a bulleted list)

I'm certain this list is not exhaustive, but hopeful it shows the algorithm complexity that's being requested.

Reply:
I think @aleventhal would agree, but I'd like his opinion, too.

Reply:
Or @cyns?

Reply:
@MelSumner 
> What if we said "if the pseudo element includes a space between it and the textual content of the current node..."

Do you mean this space?

```html
<div> text</div>
<!-- ^ -->
```

Or one of these spaces?

```css
  content: "text ";
  /*            ^ */
  content: "text " / "alt";
  /*            ^ */
  content: "text" / "alt ";
  /*                    ^ */
```

Reply:
On the call, Bryan said this should refer to whether or not the element was an `inline` or similar value, versus a `block` or similar display value. I presume that would mean if either the adjacent elements (or their pseudo when joining pseudos) had a "block-like" display (perhaps any display value other than `inline`) then use a space joiner... 

Unknown at the moment whether that is implementable, and also unknown is how far the render tree should consider "adjacent" render objects... For example, if the adjacent DOM element is `display:none`, is the expectation that the renderer check the next adjacent element or pseudo that is not `display:none`? What about a `display:inline` element with `visibility:hidden`?